pkg/syncer/spec/mutators/deployment.go

```diff
@@ -64,9 +79,49 @@ func (dm *DeploymentMutator) Mutate(downstreamObj *unstructured.Unstructured) er
 	// Alias for the template.spec to improve readability.
 	templateSpec := &deployment.Spec.Template.Spec
 
+	desiredServiceAccountName := "default"
+	if templateSpec.ServiceAccountName != "" && templateSpec.ServiceAccountName != "default" {
+		desiredServiceAccountName = templateSpec.ServiceAccountName
+	}
+
+	secretList, err := dm.secretLister.ByIndex(dm.indexName, dm.clusterName.String())
+	if err != nil {
+		return fmt.Errorf("error listing secrets for cluster %s: %w", dm.clusterName, err)
+	}
+
+	desiredSecretName := ""
+	for _, secretObj := range secretList {
+		var secret *corev1.Secret
+		switch obj := secretObj.(type) {
+		case *corev1.Secret:
+			secret = obj
+		case *unstructured.Unstructured:
+			unstructured := secretObj.(*unstructured.Unstructured).UnstructuredContent()
+			err = runtime.DefaultUnstructuredConverter.
+				FromUnstructured(unstructured, &secret)
+			if err != nil {
+				klog.Errorf("Failed to convert unstructured object to secret: %v", err)
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown object type: %T", secretObj)
+		}
+
+		// Find the SA token that matches the service account name.
+		if val, ok := secret.Annotations[corev1.ServiceAccountNameKey]; ok && val == desiredServiceAccountName {
+			desiredSecretName = PrefixSecretName(secret.Name)
+		}
+	}
+
+	if desiredSecretName == "" {
+		return fmt.Errorf("couldn't find a token upstream for the serviceaccount %s", desiredServiceAccountName)
```

Comment:
qualify with namespace or whatever else is helpful here.

Reply:
still open

Reply:
done